### PR TITLE
Attempt at getting Gdk events to get finalized and freed on the UI th…

### PR DESCRIPTION
**Pull request by Therzok**
…read

Possible fix for VSTS #739030 - Visual Studio Community 7.7 for Mac - crahses when holding down cursor key

### CODE REVIEW

File: gdk/Event.cs [hunk 1]
```diff
@@ -57,10 +65,39 @@ public Event(IntPtr raw) : this(raw, false)
 		public Event(IntPtr raw, bool owned) 
 		{
 			this.raw = raw;
+			this.owned = owned;
 			if (!owned)
 				System.GC.SuppressFinalize (this);
 		}
 
+		static object lockObject = new object ();
```
Comment (alanmcgovern) [inline gdk/Event.cs]: `static readonly` ?

File: gdk/Event.cs [hunk 2]
```diff
@@ -57,10 +65,39 @@ public Event(IntPtr raw) : this(raw, false)
...
+				gdk_event_free (@event.Handle);
+
+			return false;
+		}
+
+		public void Dispose ()
+		{
+			if (owned) {
+				System.GC.SuppressFinalize (this);
+				gdk_event_free (Handle);
```
Comment (alanmcgovern) [inline gdk/Event.cs]: zero out `Handle` to prevent crashes due to double free?

File: gdk/Event.cs [hunk 3]
```diff
@@ -57,10 +65,39 @@ public Event(IntPtr raw) : this(raw, false)
...
+			this.owned = owned;
 			if (!owned)
 				System.GC.SuppressFinalize (this);
 		}
 
+		static object lockObject = new object ();
+		static List<Event> PendingFrees = new List<Event> ();
+		static bool idleQueued;
+
+		static GLib.TimeoutHandler PerformQueuedFreesHandler = PerformQueuedFrees;
```
Comment (Therzok) [inline gdk/Event.cs]: `static readonly` here too

**Comment by Therzok** [comment]
Updated with PR feedback.

**Comment by slluis** [comment]
@alanmcgovern can you review?

**Comment by Therzok** [comment]
```
using System;
using System.IO;
using Gdk;
using Gtk;

public partial class MainWindow : Gtk.Window
{
    public MainWindow() : base(Gtk.WindowType.Toplevel)
    {
        Build();
    }

	protected override bool OnExposeEvent(EventExpose evnt)
	{
		for (int i = 0; i < 1000000; ++i)
		{
			var ev = Gtk.Application.CurrentEvent;
			var copy = new Gdk.Event(ev.Handle, true);
		}

		System.GC.Collect();
		System.GC.Collect();
		System.GC.Collect();
		System.GC.WaitForPendingFinalizers();

		return base.OnExposeEvent(evnt);
	}
}

    public class MainClass
    {
        public static void Main(string[] args)
        {
            Application.Init();

			var mw = new MainWindow();
			mw.ShowAll();

			Application.Run();

		}
    }
```

**Comment by Therzok** [comment]
Actually, this doesn't fix the issue...

The event goes out of scope and it gets freed by native gtk. I think our only solution is to make it a ref struct and avoid the value from escaping, otherwise it'll just free invalid memory.

**Comment by Therzok** [comment]
I thought it was an UI thread bug, but it can still happen.

**Comment by Therzok** [comment]
`EventOwnerChange` as it's not read-only, the rest of the events could be solved by just making a copy of the GdkEvent.

**Comment by Therzok** [comment]
Closing. The problem with Gdk.Event is how it's bound and it requires breaking API to fix. We worked around it in VSMac by manually pinvoking

**Pull request closed without merging**